File: src/typegraph/deno/src/io.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0
import { inspect } from "node:util";
// import { createInterface, Interface } from "node:readline";
import process from "node:process";
import { rpcRequest } from "./gen/client.ts";
/**
 * see: module level documentation `meta-cli/src/deploy/actors/task.rs`
 */

const JSONRPC_VERSION = "2.0";

function writeRpcMessage(message: string) {
  // split into 32-KiB chunks
  const chunkSize = 32758; // 32 KiB - 10 bytes for "jsonrpc^: " or "jsonrpc$: "
  for (let i = 0; i < message.length; i += chunkSize) {
    const chunk = message.slice(i, i + chunkSize);
    if (i + chunkSize < message.length) {
      process.stdout.write(`jsonrpc^: ${chunk}\n`);
      continue;
    }
    process.stdout.write(`jsonrpc$: ${message.slice(i, i + chunkSize)}\n`);
  }
}

type RpcNotificationMethod =
  | "Serialize"
  | "Debug"
  | "Info"
  | "Warning"
  | "Error"
  | "Success"
  | "Failure";

export const rpcNotify = (
  method: RpcNotificationMethod,
  // deno-lint-ignore no-explicit-any
  params: any = null,
): void => {
  const message = JSON.stringify({
    jsonrpc: JSONRPC_VERSION,
    method,
    params,
  });
  writeRpcMessage(message);
};

// deno-lint-ignore no-explicit-any
function getOutput(args: any[]) {
  return args
    .map((arg) => {
      if (typeof arg === "string") return arg;
      return inspect(arg, {
        colors: process.stdout.isTTY,
        depth: 10,
        maxStringLength: 1000,
        maxArrayLength: 20,
      });
    })
    .join(" ");
}

export const log: {
  // deno-lint-ignore no-explicit-any
  debug(...args: any[]): void;
  // deno-lint-ignore no-explicit-any
  info(...args: any[]): void;
  // deno-lint-ignore no-explicit-any
  warn(...args: any[]): void;
  // deno-lint-ignore no-explicit-any
  error(...args: any[]): void;
  // deno-lint-ignore no-explicit-any
  failure(data: any): void;
  // deno-lint-ignore no-explicit-any
  success(data: any, noEncode?: boolean): void;
} = {
  debug(...args): void {
    rpcNotify("Debug", { message: getOutput(args) });
  },
  info(...args): void {
    rpcNotify("Info", { message: getOutput(args) });
  },
  warn(...args): void {
    rpcNotify("Warning", { message: getOutput(args) });
  },
  error(...args): void {
    rpcNotify("Error", { message: getOutput(args) });
  },
  failure(data): void {
    rpcNotify("Failure", { data: data });
  },
  success(data, noEncode = false): void {
    if (noEncode) {
      rpcNotify("Success", { data: JSON.parse(data) });
    } else {
      rpcNotify("Success", { data: data });
    }
  },
};

export interface DeployTarget {
  baseUrl: string;
  auth: {
    username: string;
    password: string;
  };
}

export interface MigrationAction {
  apply: boolean;
  create: boolean;
  reset: boolean;
}

export interface DeployData {
  secrets: Record<string, string>;
  defaultMigrationAction: MigrationAction;
  migrationActions: Record<string, MigrationAction>;
}

export const rpc = {
  getDeployTarget: (): DeployTarget => rpcRequest("GetDeployTarget") as DeployTarget,
  getDeployData: (typegraph: string): DeployData =>
    rpcRequest<DeployData, { typegraph: string }>(
      "GetDeployData",
      { typegraph },
      false, // Don't transform object keys
    ),
};

File: src/typegraph/python/typegraph/io.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

from typing import Any, Optional, Dict
from fileinput import FileInput
from dataclasses import dataclass
from typegraph.graph.shared_types import BasicAuth
from typegraph.gen.exports.core import MigrationAction
import json


_JSONRPC_VERSION = "2.0"


def write_rpc_message(message: str):
    # we do not chunk the message as Python's print function supports long lines
    print(f"jsonrpc$: {message}")


def rpc_notify(method: str, params: Any):
    message = json.dumps(
        {
            "jsonrpc": _JSONRPC_VERSION,
            "method": method,
            "params": params,
        }
    )
    write_rpc_message(message)


class Log:
    @staticmethod
    def __format(*largs: Any):
        return " ".join(map(str, largs))

    @staticmethod
    def debug(*largs: Any):
        rpc_notify("Debug", {"message": Log.__format(*largs)})

    @staticmethod
    def info(*largs: Any):
        rpc_notify("Info", {"message": Log.__format(*largs)})

    @staticmethod
    def warn(*largs: Any):
        rpc_notify("Warning", {"message": Log.__format(*largs)})

    @staticmethod
    def error(*largs: Any):
        rpc_notify("Error", {"message": Log.__format(*largs)})

    @staticmethod
    def failure(data: Any):
        rpc_notify("Failure", {"data": data})

    @staticmethod
    def success(data: Any, noencode: bool = False):
        if noencode:
            parsed = json.loads(data)
            rpc_notify("Success", {"data": parsed})
        else:
            rpc_notify("Success", {"data": data})


class _RpcResponseReader:
    input: FileInput

    def __init__(self):
        self.input = FileInput("-")

    def read(self, rpc_id: int):
        while True:
            line = self.input.readline()
            try:
                parsed = json.loads(line)
            except Exception:
                Log.error("rpc response: failed to parse input as json")
                continue

            if parsed.get("jsonrpc") != _JSONRPC_VERSION:
                Log.error("rpc response: invalid jsonrpc version")
                continue

            if parsed.get("id") != rpc_id:
                Log.error(
                    f"rpc response: expected sequential requestests, unexpected rpc id {parsed.get('id')}"
                )
                continue

            return parsed.get("result")


class _RpcCall:
    response_reader = _RpcResponseReader()
    latest_rpc_id = 0

    @classmethod
    def call(cls, method: str, params: Any):
        cls.latest_rpc_id = cls.latest_rpc_id + 1
        rpc_id = cls.latest_rpc_id
        rpc_message = json.dumps(
            {
                "jsonrpc": _JSONRPC_VERSION,
                "id": rpc_id,
                "method": method,
                "params": params,
            }
        )

        write_rpc_message(rpc_message)
        return cls.response_reader.read(rpc_id)


@dataclass
class DeployTarget:
    base_url: str
    auth: BasicAuth


@dataclass
class DeployData:
    secrets: Dict[str, str]
    default_migration_action: MigrationAction
    migration_actions: Dict[str, MigrationAction]


def migration_action_from_dict(raw: Dict[str, bool]) -> MigrationAction:
    return MigrationAction(
        apply=raw.get("apply", False),
        create=raw.get("create", False),
        reset=raw.get("reset", False),
    )


class Rpc:
    _deploy_target: Optional[DeployTarget] = None

    # cached
    @classmethod
    def get_deploy_target(cls) -> DeployTarget:
        if cls._deploy_target is None:
            # TODO validation??
            res = _RpcCall.call("GetDeployTarget", None)

            raw_auth = res.get("auth")
            if raw_auth is None:
                raise Exception(f"invalid data from rpc call: {res}")

            auth = BasicAuth(raw_auth.get("username"), raw_auth.get("password"))

            cls._deploy_target = DeployTarget(
                base_url=res["baseUrl"],
                auth=auth,
            )

        return cls._deploy_target

    @staticmethod
    def get_deploy_data(typegraph: str) -> DeployData:
        res = _RpcCall.call("GetDeployData", {"typegraph": typegraph})

        return DeployData(
            secrets=res["secrets"],
            default_migration_action=migration_action_from_dict(
                res["defaultMigrationAction"]
            ),
            migration_actions={
                k: migration_action_from_dict(v)
                for k, v in res["migrationActions"].items()
            },
        )
